### ISSUE TITLE
Text is too small in CHT devices

### ISSUE THREAD

**Issue by roytam1** [BernardoGiordano/PKSM#654]
CHT devices(Sell officially in Hong Kong and Taiwan) has smaller font size (20x24px) than others (25x30px) for more spaces to contain larger amount characters. (Source: converted bitmap from official cbf_zh-Hant-TW.bcfnt / cbf_std.bcfnt with ctr_FontConverter)

Possible solution to deal with such situation:
check width of "あ" with fontGetCharWidthInfo, CHT devices will return smaller value. Since ideograph is square, so normally width == height. (CHECK: does it return "24" in C/J/U/E? and does it return "19" in T?)
and then adjust scale factor before print.

Cross-post:
https://github.com/fincs/new-hbmenu/issues/35
https://github.com/Steveice10/FBI/issues/423

**Comment by BernardoGiordano** [comment]
Can you send multiple screenshots of the whole user interface? Having them in a zip file attached here would be a big help

**Comment by roytam1** [comment]
sure.

![2018-01-08_23-44-42 705_top](https://user-images.githubusercontent.com/1192577/34678432-0de18ad6-f48b-11e7-970d-7dd1d2209a25.png)
![2018-01-08_23-44-42 705_top](https://user-images.githubusercontent.com/1192577/34678437-11e36352-f48b-11e7-9188-0b6a741072f7.png)

**Comment by BernardoGiordano** [comment]
the solution you proposed is feasible, I hope for it to be implemented properly in a possible rewrite, otherwise I'll put together something rough to patch it at least for now

**Comment by fincs** [comment]
For reference, this is how I fixed it in new-hbmenu: https://github.com/fincs/new-hbmenu/commit/79e2e67edf04ed2484ef9a27b8175386f71f621f

**Comment by roytam1** [comment]
Just ported my mod over @fincs 's commit to pp2d:

````
diff --git a/source/pp2d/pp2d.c b/source/pp2d/pp2d.c
index 785c4f5..e4c6612 100644
--- a/source/pp2d/pp2d.c
+++ b/source/pp2d/pp2d.c
@@ -41,6 +41,7 @@ static C3D_Mtx projectionTopLeft;
 static C3D_Mtx projectionTopRight;
 static C3D_Mtx projectionBot;
 static C3D_Tex* glyphSheets;
+static float s_textScale;
 static textVertex_s* textVtxArray;
 static int textVtxArrayPos;
 static C3D_RenderTarget* topLeft;
@@ -159,6 +160,9 @@ void pp2d_draw_text_wrap(float x, float y, float scaleX, float scaleY, u32 color
        float firstX = x;
        int lastSheet = -1;

+       scaleX *= s_textScale;
+       scaleY *= s_textScale;
+
        do
        {
                if (!*p) break;
@@ -391,6 +395,8 @@ Result pp2d_init(void)
                tex->lodParam = 0;
                #endif
        }
+       charWidthInfo_s* cwi = fontGetCharWidthInfo(fontGlyphIndexFromCodePoint(0x3042));
+       s_textScale = 20.0f / (cwi->glyphWidth); // 20 is glyphWidth in J machines

        textVtxArray = (textVertex_s*)linearAlloc(sizeof(textVertex_s)*TEXT_VTX_ARRAY_COUNT);
        C3D_BufInfo* bufInfo = C3D_GetBufInfo();
@@ -443,6 +449,9 @@ static void pp2d_get_text_size_internal(float* width, float* height, float scale
     float firstX = x;
     const uint8_t* p = (const uint8_t*)text;

+       scaleX *= s_textScale;
+       scaleY *= s_textScale;
+
     do
     {
         if (!*p) break;

````

and it looks better (Some Japanese glyphs are missing in TWN/CHT machines as expected):
![t_title_atop](https://user-images.githubusercontent.com/1192577/34915058-b76f8172-f916-11e7-910a-fd23530bc76b.png)
![t_title_bot](https://user-images.githubusercontent.com/1192577/34915059-b799fc7c-f916-11e7-8cc0-fb006ba2e78e.png)
![t_box_atop](https://user-images.githubusercontent.com/1192577/34915060-bb6f1896-f916-11e7-8c64-d47c0e0fb2bc.png)
![t_box_bot](https://user-images.githubusercontent.com/1192577/34915061-bb9b58de-f916-11e7-822b-4c8aac1dd5b1.png)

**Comment by roytam1** [comment]
> @BernardoGiordano closed this in 0445e96 5 hours ago
will you merge same change to others using pp2d, like https://github.com/BernardoGiordano/Checkpoint ?
